File: mobile_app/lib/utils/encryption.dart
import 'dart:convert';
import 'dart:typed_data';
import 'dart:math';
import 'package:encrypt/encrypt.dart' as encrypt;

/// Shared AES-256 key (must match backend and frontend).
/// In production, this should come from secure storage or env vars.
const String _symmetricKey = 'xJSJGB6vtXCjeljCNQ5hD0GeQXDMRtQ/ZjMFifrYBb8=';

final _key = encrypt.Key.fromBase64(_symmetricKey);

/// Encrypt a JSON-serializable Map into a Base64 string with random IV prepended.
String encryptPayload(Map<String, dynamic> data) {
  final jsonStr = jsonEncode(data);
  final iv = encrypt.IV.fromSecureRandom(16);

  final encrypter = encrypt.Encrypter(
    encrypt.AES(_key, mode: encrypt.AESMode.cbc, padding: 'PKCS7'),
  );
  final encrypted = encrypter.encryptBytes(utf8.encode(jsonStr), iv: iv);

  // Combine IV + Ciphertext
  final combined = Uint8List.fromList(iv.bytes + encrypted.bytes);
  return base64.encode(combined);
}

/// Decrypt a Base64 string (with prepended IV) into a Map.
Map<String, dynamic> decryptPayload(String encryptedStr) {
  // Strip surrounding quotes if present (backend wraps in JSON quotes)
  final cleanStr = encryptedStr.replaceAll(RegExp(r'^"|"$'), '');
  final rawBytes = base64.decode(cleanStr);

  if (rawBytes.length < 16) {
    throw Exception('Encrypted data too short');
  }

  final iv = encrypt.IV(Uint8List.fromList(rawBytes.sublist(0, 16)));
  final ciphertext = encrypt.Encrypted(Uint8List.fromList(rawBytes.sublist(16)));

  final encrypter = encrypt.Encrypter(
    encrypt.AES(_key, mode: encrypt.AESMode.cbc, padding: 'PKCS7'),
  );

  final decryptedStr = encrypter.decrypt(ciphertext, iv: iv);
  return jsonDecode(decryptedStr) as Map<String, dynamic>;
}

/// Decrypt a response body string into a dynamic value (can be List or Map).
dynamic decryptResponseBody(String responseBody) {
  // Strip surrounding quotes if present
  final cleanStr = responseBody.replaceAll(RegExp(r'^"|"$'), '');
  final rawBytes = base64.decode(cleanStr);

  if (rawBytes.length < 16) {
    throw Exception('Encrypted data too short');
  }

  final iv = encrypt.IV(Uint8List.fromList(rawBytes.sublist(0, 16)));
  final ciphertext = encrypt.Encrypted(Uint8List.fromList(rawBytes.sublist(16)));

  final encrypter = encrypt.Encrypter(
    encrypt.AES(_key, mode: encrypt.AESMode.cbc, padding: 'PKCS7'),
  );

  final decryptedStr = encrypter.decrypt(ciphertext, iv: iv);
  return jsonDecode(decryptedStr);
}

File: frontend/src/utils/encryption.ts
import CryptoJS from 'crypto-js';

// Shared key must be exactly 32 bytes for AES-256
// This should ideally come from an environment variable in a real app
const SYMMETRIC_KEY = "xJSJGB6vtXCjeljCNQ5hD0GeQXDMRtQ/ZjMFifrYBb8=";
const key = CryptoJS.enc.Base64.parse(SYMMETRIC_KEY);

/**
 * Encrypt a JSON object into a Base64 string with random IV prepended.
 */
export const encryptPayload = (data: any): string => {
  const jsonStr = JSON.stringify(data);
  const iv = CryptoJS.lib.WordArray.random(16);
  
  const encrypted = CryptoJS.AES.encrypt(jsonStr, key, {
    iv: iv,
    mode: CryptoJS.mode.CBC,
    padding: CryptoJS.pad.Pkcs7
  });
  
  // Combine IV + Ciphertext
  const ivAndCiphertext = iv.concat(encrypted.ciphertext);
  return CryptoJS.enc.Base64.stringify(ivAndCiphertext);
};

/**
 * Decrypt a Base64 string (with prepended IV) into a JSON object.
 */
export const decryptPayload = (encryptedStr: string): any => {
  try {
    const rawData = CryptoJS.enc.Base64.parse(encryptedStr);
    
    // Extract 16-byte IV
    const iv = CryptoJS.lib.WordArray.create(rawData.words.slice(0, 4));
    // Extract Ciphertext
    const ciphertext = CryptoJS.lib.WordArray.create(rawData.words.slice(4));
    
    const decrypted = CryptoJS.AES.decrypt(
      { ciphertext: ciphertext } as any,
      key,
      {
        iv: iv,
        mode: CryptoJS.mode.CBC,
        padding: CryptoJS.pad.Pkcs7
      }
    );
    
    const decryptedStr = decrypted.toString(CryptoJS.enc.Utf8);
    return JSON.parse(decryptedStr);
  } catch (error) {
    console.error("Decryption failed:", error);
    throw new Error("Failed to decrypt payload");
  }
};
